cpp/src/arrow/util/decimal.h

```diff
@@ -31,6 +31,263 @@
+ public:
+  /// \cond FALSE
+  // (need to avoid a duplicate definition in sphinx)
+  using BasicDecimal32::BasicDecimal32;
+  /// \endcond
+
+  /// \brief constructor creates a Decimal32 from a BasicDecimal32
+  constexpr Decimal32(const BasicDecimal32& value) noexcept  // NOLINT runtime/explicit
+      : BasicDecimal32(value) {}
+
+  /// \brief Parse the number from a base 10 string representation
+  explicit Decimal32(const std::string& value);
+
+  /// \brief Empty constructor creates a Decimal32 with a value of 0
+  /// this is required for some older compilers
+  constexpr Decimal32() noexcept : BasicDecimal32() {}
+
+  /// \brief Divide this number by right and return the result.
+  ///
+  /// This operation is not destructive.
+  /// The answer rounds to zero. Signs work like:
+  ///   21 /  5 ->  4,  1
+  ///  -21 /  5 -> -4, -1
+  ///   21 / -5 -> -4,  1
+  ///  -21 / -5 ->  4, -1
+  /// \param[in] divisor the number to divide by
+  /// \return the pair of the quotient and the remainder
+  Result<std::pair<Decimal32, Decimal32>> Divide(const Decimal32& divisor) const {
+    std::pair<Decimal32, Decimal32> result;
+    auto dstatus = BasicDecimal32::Divide(divisor, &result.first, &result.second);
+    ARROW_RETURN_NOT_OK(ToArrowStatus(dstatus));
+    return result;
+  }
+
+  /// \brief Convert the Decimal32 value to a base 10 decimal string with the given scale
+  std::string ToString(int32_t scale) const;
+
+  /// \brief Convert the value to an integer string
+  std::string ToIntegerString() const;
+
+  /// \brief Cast this value to an int64_t
+  explicit operator int64_t() const;
+
+  explicit operator Decimal64() const;
+
+  /// \brief Convert a decimal string to a Decimal value, optionally including
+  /// precision and scale if they're passed in and not null.
+  static Status FromString(std::string_view s, Decimal32* out, int32_t* precision,
+                           int32_t* scale = NULLPTR);
+  static Status FromString(const std::string& s, Decimal32* out, int32_t* precision,
+                           int32_t* scale = NULLPTR);
+  static Status FromString(const char* s, Decimal32* out, int32_t* precision,
+                           int32_t* scale = NULLPTR);
+  static Result<Decimal32> FromString(std::string_view s);
+  static Result<Decimal32> FromString(const std::string& s);
+  static Result<Decimal32> FromString(const char* s);
+
+  static Result<Decimal32> FromReal(double real, int32_t precision, int32_t scale);
+  static Result<Decimal32> FromReal(float real, int32_t precision, int32_t scale);
+
+  /// \brief Convert from a big-endian byte representation. The length must be
+  ///        between 1 and 4
+  /// \return error statis if the length is an invalid value
+  static Result<Decimal32> FromBigEndian(const uint8_t* data, int32_t length);
+
+  static void SwapEndian(const uint32_t* old_data, uint32_t* const new_data,
+                         const int64_t length);
```

Comment:
This makes the usage in util.cc a bit nicer, but as a method on Decimal itself it might cause confusion. I was imagining
```suggestion
  Decimal32 SwapEndian() const;
```

Then util.cc would look like
```c++
  template <typename T>
  enable_if_decimal<T, Status> Visit(const T& type) {
    using value_type = typename TypeTraits<T>::CType;
    auto data = data_->buffers[1]->span_as<value_type>();
    auto new_data = new_buffer->mutable_data_as<value_type>();
    for (value_type v : data) {
      *new_data++ = v.SwapEndian();
    }
```

... actually now that I think of it, it'd also be pretty straightforward to just use ToBytes instead of adding SwapEndian:
```c++
  template <typename T>
  enable_if_decimal<T, Status> Visit(const T& type) {
    using value_type = typename TypeTraits<T>::CType;
    auto data = data_->buffers[1]->span_as<value_type>();
    auto new_data = new_buffer->mutable_data_as<value_type>();
    for (value_type v : data) {
      auto bytes = v.ToBytes();
      std::reverse(bytes.begin(), bytes.end());
      memcpy(new_data++, &bytes, sizeof(bytes));
    }
```

Reply:
updated using the suggestion, let me know what you think